### PR TITLE
feat(multicluster): add weight to RemoteCluster

### PR DESCRIPTION
**Pull request by allenlsy**
Signed-off-by: Allen Leigh <allenlsy@gmail.com>

**Description**:

Add `weight` attribute to remote cluster spec. This PR prepares for implementing locality based routing. 

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | X |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

No

1. Is this a breaking change?
No

### CODE REVIEW

File: charts/osm/crds/multiclusterservice.yaml [hunk 1]
```diff
@@ -58,6 +58,9 @@ spec:
                       name:
                         description: Name of the remote cluster
                         type: string
+                      weight:
+                        description: Load balancing weight of the remote cluster in percentage
+                        type: integer
```
Comment (snehachhabria) [inline charts/osm/crds/multiclusterservice.yaml:63]: should we add a check to ensure its a positive integer and if not specified assign a default of 0 ?

Comment (allenlsy) [reply charts/osm/crds/multiclusterservice.yaml:63]: One correction I added: the weight here is not percentage.

File: charts/osm/crds/multiclusterservice.yaml [hunk 2]
```diff
@@ -58,6 +58,12 @@ spec:
                       name:
                         description: Name of the remote cluster
                         type: string
+                      weight:
+                        description: Load balancing weight of the remote cluster
+                        type: integer
+                        minimum: 1
```
Comment (snehachhabria) [inline charts/osm/crds/multiclusterservice.yaml:64]: this doesn't seem right, the minimum should be 0

File: charts/osm/crds/multiclusterservice.yaml [hunk 3]
```diff
@@ -58,6 +58,12 @@ spec:
...
+                        maximum: 10000
```
Comment (snehachhabria) [inline charts/osm/crds/multiclusterservice.yaml:65]: Any particular reason why this is capped at 10000 ?

File: charts/osm/crds/multiclusterservice.yaml (hunk 2 above)
Comment (allenlsy) [reply charts/osm/crds/multiclusterservice.yaml:64]: 0 means no weight at all. If that is the case, why we add that cluster?

File: charts/osm/crds/multiclusterservice.yaml (hunk 3 above)
Comment (allenlsy) [reply charts/osm/crds/multiclusterservice.yaml:65]: No, I confused myself with another number.

Anyway, in Envoy, there is no limit on the load balancing weight. If we follow convention, we can do the same. https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/endpoint/v3/endpoint_components.proto#envoy-v3-api-field-config-endpoint-v3-localitylbendpoints-load-balancing-weight

File: charts/osm/crds/multiclusterservice.yaml (hunk 2 above)
Comment (snehachhabria) [reply charts/osm/crds/multiclusterservice.yaml:64]: yes, there can be scenarios where the cluster is mentioned but there's no traffic yet

File: charts/osm/crds/multiclusterservice.yaml (hunk 1 above)
Comment (allenlsy) [reply charts/osm/crds/multiclusterservice.yaml:63]: Updated minimum to 0, and no limit on maximum, as long as it's in the range of integer

**Review by ksubrmnn — APPROVED** [review]

File: pkg/apis/config/v1alpha1/multi_cluster_service.go [hunk 4]
```diff
@@ -39,6 +39,12 @@ type ClusterSpec struct {
 
 	// Name defines the name of the remote cluster.
 	Name string `json:"name,omitempty"`
+
+	// Weight defines the load balancing weight of the remote cluster
+	Weight int `json:"weight,omitempty"`
+
+	// Certificate defines the mTLS certificate of the remote cluster
+	Certificate string `json:"certificate,omitempty"`
```
Comment (shashankram) [inline pkg/apis/config/v1alpha1/multi_cluster_service.go:47]: This PR does not mention anything regarding certificates. Please drop this change or describe in the commit why this is required.

**Review by shashankram — CHANGES_REQUESTED** [review]

**Review by snehachhabria — CHANGES_REQUESTED** [review]
Agree with @shashankram lets please rebase based off the latest multicluster changes and determine if this is needed. Also for any changes in multicluter, going forward lets have a proposal and get it reviewed. Post the proposal approval we can work on a POC to validate the proposal before we commit changes to the repository

File: pkg/apis/config/v1alpha1/multi_cluster_service.go (hunk 4 above)
Comment (allenlsy) [reply pkg/apis/config/v1alpha1/multi_cluster_service.go:47]: Certificate field was originally in the CRD, but not in the Go struct. I think this was a mistake for leaving it out. Thus adding it back in this PR.

Comment (shashankram) [reply pkg/apis/config/v1alpha1/multi_cluster_service.go:47]: Where is the certificate being used? Is including it in the CRD necessary? Like I mentioned already, the commit message does not talk about the certificate, so seeing this here is quite confusing.

Comment (allenlsy) [reply pkg/apis/config/v1alpha1/multi_cluster_service.go:47]: I haven't checked if the `certificate` is used anywhere in the code base. If we find it's not needed, we can remove it in another PR.

Comment (shashankram) [reply pkg/apis/config/v1alpha1/multi_cluster_service.go:47]: If you aren't sure it's needed, there's no reason to include it in this change.

Comment (snehachhabria) [reply pkg/apis/config/v1alpha1/multi_cluster_service.go:47]: the crd is slate, the current crd doesn't have `certificate`. Please rebase changes and share a proposal as to why weight needs to be added for the remote cluster

**Review by snehachhabria — APPROVED** [review]

**Pull request merged**